### PR TITLE
Block P/Invokes with UnmanagedCallersOnlyAttribute

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
See https://github.com/dotnet/runtime/issues/38209#issuecomment-650576664 regarding mono implementation.

/cc @jkotas @lambdageek

### CODE REVIEW

**Comment by lambdageek** [comment]
@AaronRobinsonMSFT there's no mono implementation yet. :grin:   I just had an idea this morning and thought it was worth writing down.

**Comment by AaronRobinsonMSFT** [comment]
@lambdageek Ah. Someone pointed out your comment offline. Once I realized it wasn't handled in coreclr I decided to investigate. Took very little to actually validate and see the problem. This seems more like an optimization rather than anything that we _need_ to support, but after talking it out I realized it does have value and I don't think too much risk. Let me know if you think it is worth adding or blocking.

**Comment by jkotas** [comment]
> This seems more like an optimization

I do not think it is just an optimization. It affects the function pointer behavior you get from these in observable way. Before this change, the function pointer was managed. After this change, the function pointer is unmanaged.

There are other places in the runtime that get method entrypoint and that do not go through `TryGetMultiCallableAddrOfCode`. They will still get the old behavior.

Do we need a test for this? E.g. what happens when you create delegate to one of these methods and call that delegate; what happens when you call this method via reflection; etc.

I am wondering whether it would be best to error out for this case.

**Comment by AaronRobinsonMSFT** [comment]
I meant scenario optimization, not performance. Perhaps "convenience semantic" is a more appropriate description. The only reason I can see doing this is to permit the ability to reuse a variable/member. For example, a variable/member is initialized with a function pointer that was constructed by a `DllImport` and then later on the variable/member could be updated by the result of the [`NativeLibrary.GetExport()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary.getexport) API. The developer could then use a C# function pointer variable/member regardless of origin. This is what I meant by "optimization".

Your statement about the result of the `&` is indeed correct. From a users perspective I think applying the semantics of `UnmanagedCallersOnlyAttribute` only as the result of `&` is unusual but could be easily described if one considers the intent of the two attributes. The combination of `UnmanagedCallersOnlyAttribute` and `DllImport` seems weird in general except for the previously state use case.

>There are other places in the runtime that get method entrypoint and that do not go through TryGetMultiCallableAddrOfCode. They will still get the old behavior.

>Do we need a test for this? E.g. what happens when you create delegate to one of these methods and call that delegate; what happens when you call this method via reflection; etc.

Yes, tests are warranted. This PR isn't complete since I didn't add tests and figured I missed something - as alluded to above. What other entry points? I looked through layers and honestly it was a bit tough to navigate. The guts of `TryGetMultiCallableAddrOfCode` are a bit opaque to say the least. Suggestions on how to trigger these code paths are appreciated.

>I am wondering whether it would be best to error out for this case.

That is another reason I didn't want to write more tests. Without function pointers in the repo these tests get a bit wordy. However, I assume that all the places I would need to make this work are the same places I would need to throw an exception so a few simple tests will be needed regardless.

**Comment by jkotas** [comment]
> What other entry-points?

E.g. `MethodDesc::GetSingleCallableAddrOfCode`. `MethodDesc::GetSingle/MultipleCallableAddrOfCode` and other similar entrypoint API are meant to just produce a function pointer that one can call, without doing any heavy lifting to produce the code that is behind the function pointer. For example, these APIs won't do any JITing today.

Checks like this may be better to add to the existing places where we are doing the heavy lifting, e.g. where we are generating or JITing the marshalling stubs.

File: src/coreclr/src/vm/dllimport.cpp [hunk 1]
```diff
@@ -3046,9 +3046,21 @@ BOOL NDirect::MarshalingRequired(MethodDesc *pMD, PCCOR_SIGNATURE pSig /*= NULL*
...
-        if (pMD->IsNDirect() && ((NDirectMethodDesc *)pMD)->IsClassConstructorTriggeredByILStub())
-            return TRUE;
+        if (pMD->IsNDirect())
+        {
+            // A P/Invoke marked with UnmanagedCallersOnlyAttribute
+            // doesn't technically require marshalling. However, we
+            // don't support a DllImport with this attribute and we
+            // error out during IL Stub generation so we indicate that
+            // when checking if an IL Stub is needed.
+            if (pMD->HasUnmanagedCallersOnlyAttribute())
```
Comment (jkotas) [inline src/coreclr/src/vm/dllimport.cpp:3056]: The same check should be also in `IsMarshallingRequired(MethodDesc targetMethod)` in `src\coreclr\src\tools\aot\ILCompiler.ReadyToRun\Interop\IL\Marshaller.ReadyToRun.cs`

File: src/coreclr/src/vm/prestub.cpp [hunk 2]
```diff
@@ -2169,6 +2169,11 @@ PCODE MethodDesc::DoPrestub(MethodTable *pDispatchingMT, CallerGCMode callerGCMo
             }
         }
 
+        // P/Invoke marked with UnmanagedCallersOnlyAttribute is not
+        // presently supported.
+        if (HasUnmanagedCallersOnlyAttribute())
```
Comment (jkotas) [inline src/coreclr/src/vm/prestub.cpp:2174]: CreateNDirectStubWorker may be a better place for this

Comment (jkotas) [reply src/coreclr/src/vm/prestub.cpp:2174]: (It is where other similar checks for invalid signatures, etc. happen.)

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/vm/prestub.cpp:2174]: I figured putting this in the PreStub was the best place to catch all uses. Am I handling more places by putting it lower in the stack?

Comment (jkotas) [reply src/coreclr/src/vm/prestub.cpp:2174]: Putting this in PreStub is penalizing the crossgen case. For crossgen, we want to just lookup the code, without doing any error checking. If we find good pregenerated code, we know that the error checks succeeded during crossgen, no need to repeat them.

File: src/coreclr/src/vm/dllimport.cpp [hunk 3]
```diff
@@ -4179,15 +4191,21 @@ static void CreateNDirectStubAccessMetadata(
...
+    // Check if we have a MethodDesc to query for additional data.
     if (pSigDesc->m_pMD != NULL)
     {
-        lcidArg = GetLCIDParameterIndex(pSigDesc->m_pMD);
+        MethodDesc* pMD = pSigDesc->m_pMD;
+
+        // P/Invoke marked with UnmanagedCallersOnlyAttribute is not
+        // presently supported.
+        if (pMD->HasUnmanagedCallersOnlyAttribute())
+            COMPlusThrow(kNotSupportedException, IDS_EE_NDIRECT_UNSUPPORTED_SIG);
```
Comment (jkotas) [inline src/coreclr/src/vm/dllimport.cpp:4203]: Could you please also add the same check to `PInvokeILEmitter.EmitIL` in `src\coreclr\src\tools\aot\ILCompiler.ReadyToRun\IL\Stubs\PInvokeILEmitter.cs` ?

**Review by jkotas — APPROVED** [review]

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/IL/Stubs/PInvokeILEmitter.cs [hunk 4]
```diff
@@ -29,6 +29,12 @@ private PInvokeILEmitter(MethodDesc targetMethod)
         {
             Debug.Assert(targetMethod.IsPInvoke);
             _targetMethod = targetMethod;
+
+            // P/Invoke marked with UnmanagedCallersOnlyAttribute is not
```
Comment (jkotas) [inline src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/IL/Stubs/PInvokeILEmitter.cs:33]: The other similar NotSupported checks are in EmitIL. Is there a reason why this check is here and not in EmitIL?

(just would like to know whether it is intentional or not)

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/IL/Stubs/PInvokeILEmitter.cs:33]: It was originally intentionally because I was going to handle it in a similar manner as the native side (i.e. during the reading of metadata). Looking at this now, I see what you are saying and since I didn't fold the check into `GetPInvokeMethodMetadata()` I think it should be there. Will move.
